### PR TITLE
[IMP] payment(_stripe), website_sale(_delivery): support express checkout

### PR DESCRIPTION
**Pull request by chevalierv**
This commit enables eCommerce customers to pay with the express payment
methods Apple Pay and Google Pay from the cart page.

For the moment, only Stripe supports this additional feature but it
was designed to make it easy to implement with a new provider.

task-2754209

See also:
- https://github.com/odoo/enterprise/pull/29915
- https://github.com/odoo/documentation/pull/2392

### CODE REVIEW

File: addons/website_payment_stripe/controllers/main.py
Comment (chevalierv) [inline addons/website_payment_stripe/controllers/main.py:19]: Hello @odoo/rd-security :wave: 

This PR is not yet ready for review but I wanted to ask what you advise in that case.

A bit of context: This PR allows skipping all the steps of the checkout flow from the cart to the confirmation page by using an express checkout button 'Apple Pay' or 'Google Pay'. Stripe processes the Payment. For Apple Pay to work, the customer needs to register his domain name. Stripe handles that through an API but calls back our server to check if the URL works. The URL called by Stripe needs to return the string contained in [apple-developer-merchantid-domain-association](https://github.com/odoo/odoo/blob/ccd7f8fee5ad460d7d2400e62b34f91166b7350c/addons/website_payment_stripe/static/apple-developer-merchantid-domain-association).

Is there a better (a more secure?) way than the `open()` method to do that?

See also: https://stripe.com/docs/stripe-js/elements/payment-request-button?html-or-react=html#verifying-your-domain-with-apple-pay

Have a nice weekend :sun_with_face:

File: addons/payment_stripe/static/src/js/payment_form.js
Comment (xmo-odoo) [inline addons/payment_stripe/static/src/js/payment_form.js:73]: This is an addition in master using arrow function and const, why is it using callbacks and var?

File: addons/website_payment_stripe/controllers/main.py
Comment (xmo-odoo) [reply addons/website_payment_stripe/controllers/main.py:19]: > Is there a better (a more secure?) way than the `open()` method to do that?

Yes, since you're fetching from module static data you should use the `file_open` helper function, ideally with an extension filter (though the filter feature doesn't work with no extension so not an option here).

Incidentally using `file_open` means you don't need to use `get_resource_path`, it does the normalisation internally, so you can just give it the path `"/website_payment_stripe/static/apple-developer-merchantid-domain-association"` and it should

File: addons/payment_stripe/static/src/js/payment_form.js
Comment (xmo-odoo) [inline addons/payment_stripe/static/src/js/payment_form.js:125]: It's a bit weird to create an empty object then start filling it with literal string keys, why not create the object with the two keys directly?

File: addons/payment_stripe/static/src/js/payment_form.js
Comment (xmo-odoo) [inline addons/payment_stripe/static/src/js/payment_form.js:134]: Having a local `billing_address` variable would make this a lot more readable.

File: addons/payment_stripe/static/src/js/payment_form.js
Comment (xmo-odoo) [inline addons/payment_stripe/static/src/js/payment_form.js:164]: If you're using arrow functions above, use them everywhere. `function` expressions would be useful if you named them but that's not even the case.

File: addons/website_sale/controllers/main.py
Comment (xmo-odoo) [inline addons/website_sale/controllers/main.py:926]: Nested function seems wholly unnecessary, can use a utility method (on the controller object or the model, doesn't really matter).

File: addons/website_sale/controllers/main.py
Comment (xmo-odoo) [inline addons/website_sale/controllers/main.py:947]: Using a guard form (flipping the condition) allows for quickly discarding the error path, and less nested code:

```suggestion
            if billing_errors:
                billing_errors['error_message'] = billing_error_msg
                raise ValidationError(billing_errors)

            # We need to set billing partner first to be able to save shipping partner.
            order.partner_id = billing_partner_id
            # This is the *only* thing that the front end user will see/edit anyway when
            # choosing billing address.
            order.partner_invoice_id = billing_partner_id
            order.message_partner_ids = [Command.set([billing_partner_id])]
```

File: addons/website_sale/controllers/main.py
Comment (xmo-odoo) [inline addons/website_sale/controllers/main.py:965]: Since the route is public and this is `sudo`-ed, would it not be possible for anyone to just update the information of arbitrary partners here?

File: addons/website_sale/controllers/main.py
Comment (xmo-odoo) [inline addons/website_sale/controllers/main.py:977]: Even if this is a controller object, avoid the use of "public" method forms, it's too easy to decide to e.g. move this to a model and forget to rename it.

File: addons/payment_stripe/static/src/js/payment_form.js
Comment (chevalierv) [reply addons/payment_stripe/static/src/js/payment_form.js:73]: Actually, it is a work in progress and I'm not really at ease with javascript :sweat_smile: Thanks for this remark, I'll look into it :)

**Review by xmo-odoo — CHANGES_REQUESTED** [review]
A few notes while quickly going over the code, this is by no means complete.

**Comment by stefanorigano** [comment]
hello @chevalierv , I pushed 74c97f50fd2485c05b4fbdefdedc5224cfbb6d5c on top 👋

File: addons/payment_stripe/controllers/main.py
Comment (Feyensv) [inline addons/payment_stripe/controllers/main.py:28]:
```suggestion
    _get_express_rendrering_values = '/payment/stripe/express_rendering_value'
```
:eyes:

File: addons/payment_stripe/static/src/js/payment_form.js
Comment (chevalierv) [reply addons/payment_stripe/static/src/js/payment_form.js]: Indeed, but those data are given by Stripe's javascript framework and we can't change the output.

File: addons/payment_stripe/models/payment_acquirer.py
Comment (Williambraecky) [inline addons/payment_stripe/models/payment_acquirer.py:214]:
```suggestion
                    "that you are using live credentials."
```

File: addons/payment_stripe/static/src/js/express_checkout_form.js
Comment (Williambraecky) [inline addons/payment_stripe/static/src/js/express_checkout_form.js:142]: Do we not do anything on error?

File: addons/payment_stripe/static/src/js/express_checkout_form.js
Comment (Williambraecky) [inline addons/payment_stripe/static/src/js/express_checkout_form.js:213]: TODO left here

File: addons/website_sale/controllers/main.py
Comment (Williambraecky) [inline addons/website_sale/controllers/main.py:1119]: You can unpack `billing_address` directly in the function here, it will fail when calling it without `billing_address` but it's already the case:
```suggestion
    def process_express_checkout(self, billing_address, **post):
```

File: addons/website_sale/controllers/main.py
Comment (Williambraecky) [inline addons/website_sale/controllers/main.py:1131]:
```suggestion
```

File: addons/website_sale/controllers/main.py
Comment (Williambraecky) [inline addons/website_sale/controllers/main.py:1172]: What happens if we don't have any records matching country/state?

File: addons/website_sale/controllers/main.py
Comment (Williambraecky) [inline addons/website_sale/controllers/main.py:1142]: It's a detail but this does the same thing:
```suggestion
            order.message_partner_ids = billing_partner_id
```

File: addons/website_sale/controllers/main.py
Comment (Williambraecky) [inline addons/website_sale/controllers/main.py:1155]: Linting
```suggestion
                if all(billing_address[k] == billing_partner_sudo[k] for k in billing_address):
```

File: addons/website_sale/controllers/main.py
Comment (Williambraecky) [inline addons/website_sale/controllers/main.py:1226]: I am not sure if it is the case here, but some fields on Partner should be prevented from being written to in some cases, see https://github.com/odoo/odoo/blob/master/addons/account/controllers/portal.py#L130

Comment (Williambraecky) [reply addons/website_sale/controllers/main.py:1226]: Note: it is possible to do with some other `website_sale`'s routes. I got aware of that issue only recently.

File: addons/website_sale_delivery/static/src/js/express_checkout_form.js
Comment (Williambraecky) [inline addons/website_sale_delivery/static/src/js/express_checkout_form.js:15]: I don't know how feasible it is but we don't need shipping information if the order only contains services

File: addons/payment/static/src/js/express_checkout_form.js
Comment (Williambraecky) [inline addons/payment/static/src/js/express_checkout_form.js:1]: For new js code please use ES6 like module declaration
```suggestion
/** @odoo-module **/
```

**Review by Williambraecky — COMMENTED** [review]
First initial review, mostly just read through it, thanks for your work.
I cannot assess the validity of the express checkout flow as I am not knowledgeable about it.

Those are some very small comments but the code looks very sane to me for the most part.

File: addons/payment_stripe/controllers/main.py
Comment (mart-e) [inline addons/payment_stripe/controllers/main.py:163]: does it need to be a file? can't we have simply a global variable?

**Review by mart-e — COMMENTED** [review]
@robodoo override=ci/security
Reason: comments

File: addons/website_sale_delivery/static/src/js/express_checkout_form.js
Comment (chevalierv) [reply addons/website_sale_delivery/static/src/js/express_checkout_form.js:15]: is it possible to know whether the product is a service or not from js actually or not? If so, I'll add a parameter in the method to take that into account

Comment (Williambraecky) [reply addons/website_sale_delivery/static/src/js/express_checkout_form.js:15]: We could deduce it from the page, if there are any delivery methods or not.
It's how it is done currently.. Though if you have a better idea I'm all ears

File: addons/payment_stripe/static/src/js/express_checkout_form.js
Comment (chevalierv) [reply addons/payment_stripe/static/src/js/express_checkout_form.js:213]: I know, I want to check with @AntoineVDV so I purposely left it here ^^ It will be removed before merging. Thanks for noticing it :)

File: addons/website_sale/controllers/main.py
Comment (chevalierv) [reply addons/website_sale/controllers/main.py:1172]: The record included is empty and the partner has no country/state. No error is thrown nor logged. It's unlikely to happen because the country and the state are a selection in the Express checkout form managed by Google and Apple, and the records in Odoo can't be deleted or archived.

File: addons/website_sale/controllers/main.py
Comment (chevalierv) [reply addons/website_sale/controllers/main.py:1226]: Thanks for pointing that out. I hadn't thought of it that way. I'll change the code to sanitize the fields I'm writing on.

File: addons/payment_stripe/controllers/main.py
Comment (chevalierv) [reply addons/payment_stripe/controllers/main.py:163]: We could use a global variable but the string is 9115 characters long so I think that it will make the code less readable. Stripe gave it as a file probably for the exact same reason.

File: addons/website_sale_delivery/static/src/js/express_checkout_form.js
Comment (chevalierv) [reply addons/website_sale_delivery/static/src/js/express_checkout_form.js]: I don't use the same logic to compute the delivery methods over express checkout so I can't assume that we don't need a shipping address from that. The only way to do it would be to know in js if the order is service only or not and to give the args to the override but I think it's a bit overkill for this feature.

File: addons/payment_stripe/static/src/js/express_checkout_form.js
Comment (chevalierv) [reply addons/payment_stripe/static/src/js/express_checkout_form.js]: Currently, we don't indeed. It's not a big deal since the error is already dealt with by Apple Pay and Google Pay, the user just have to open again the form. I won't merge the comments like this if we decide not to handle those errors.

File: addons/website_sale_delivery/static/src/js/express_checkout_form.js
Comment (Williambraecky) [reply addons/website_sale_delivery/static/src/js/express_checkout_form.js]: It would be good to note it for the future.
Created: task 2965823

**Review by Williambraecky — APPROVED** [review]
Code ok for me 👍 , just some minor details
Nice job

File: addons/payment/static/src/js/express_checkout_form.js
Comment (chevalierv) [reply addons/payment/static/src/js/express_checkout_form.js:28]: Thanks for the suggestion! Indeed `let self = this` was not necessary for the arrow function but for using inside a Jquery method. I implemented your changes and it works like a charm :)

File: addons/website_sale_delivery/controllers/main.py
Comment (Demesmaeker) [inline addons/website_sale_delivery/controllers/main.py:144]:
```suggestion
        """ Processes shipping address and returns available carriers
```

File: addons/website_sale_delivery/controllers/main.py
Comment (Demesmaeker) [inline addons/website_sale_delivery/controllers/main.py:146]:
```suggestion
        Depending on either the partner is registered or logged in, creates a new partner or uses
```

File: addons/website_sale_delivery/controllers/main.py
Comment (Demesmaeker) [inline addons/website_sale_delivery/controllers/main.py:153]:
```suggestion
        order_sudo = request.website.sale_get_order()
```

File: addons/website_sale_delivery/controllers/main.py
Comment (Demesmaeker) [inline addons/website_sale_delivery/controllers/main.py:217]: wrong param name

File: addons/website_sale_delivery/controllers/main.py
Comment (Demesmaeker) [inline addons/website_sale_delivery/controllers/main.py:88]: missing params

File: addons/website_sale_delivery/controllers/main.py
Comment (Demesmaeker) [inline addons/website_sale_delivery/controllers/main.py:190]: Duplicate code, might be interesting to factorize? :thinking:

File: addons/website_sale_delivery/controllers/main.py
Comment (Demesmaeker) [inline addons/website_sale_delivery/controllers/main.py:212]: Might be useful to tag it as a static method for your tests?

**Review by Demesmaeker — COMMENTED** [review]
Mostly nitpicking to ease anv review :duck: 
Nice PR overall, was interesting to read ^^

File: addons/website_sale_delivery/controllers/main.py
Comment (chevalierv) [reply addons/website_sale_delivery/controllers/main.py:212]: PRO: No need to instantiate objects in tests

File: addons/payment_stripe/models/payment_acquirer.py
Comment (chevalierv) [reply addons/payment_stripe/models/payment_acquirer.py:45]: To discuss with @pierre-pde 

And move Apple Pay register domain to the configuration tab

File: addons/payment/models/payment_acquirer.py
Comment (Williambraecky) [reply addons/payment/models/payment_acquirer.py:82]: Might as well add the comma then.

File: addons/payment_stripe/models/payment_acquirer.py
Comment (Demesmaeker) [reply addons/payment_stripe/models/payment_acquirer.py:45]: Hm, thinking of it this morning, 2 buttons'd be overkill too, as is the `registered_domain_name` field, this is literally the `base_url`, why should you store it again?

One action button can stay that'll either send the base url/check the url received from Stripe is the base_url (or you check if it's okay from Stripe, if not, you send the base url). But these fields are useless IMHO.

As for the visual aspect on the configuration/credentials page, I think that just having the button with "Verify Apple Pay compatibility" or something like this would be enough, with a rainbowman effect when everything is ok, and a userError if not.

File: addons/payment/models/payment_acquirer.py
Comment (Demesmaeker) [reply addons/payment/models/payment_acquirer.py:82]: Ha, yes, indeed ^^ Thanks!

File: addons/payment_stripe/data/payment_acquirer_data.xml
Comment (chevalierv) [reply addons/payment_stripe/data/payment_acquirer_data.xml:7]: @pierre-pde says that it's better to have it enabled by default since it will work for Google Pay. Only Apple Pay is affected by the extra configuration and at worst, only Apple Pay won't work and users will just have to follow the documentation.

File: addons/payment_stripe/models/payment_transaction.py
Comment (chevalierv) [reply addons/payment_stripe/models/payment_transaction.py:245]: I'd rather use `Whether` as it's already used in all payment functions' docstring when talking about boolean but let's see what @AntoineVDV thinks about that.

Comment (Demesmaeker) [reply addons/payment_stripe/models/payment_transaction.py:245]: I think that's a long-lost debate for me between us. "Either" for important choices, "whether" for unimportant ones, but... well... :shrug: At least I tried! :laughing:

File: addons/payment_stripe/views/payment_views.xml
Comment (Demesmaeker) [inline addons/payment_stripe/views/payment_views.xml:42]:
```suggestion
            <group name='payment_form' position="inside">
            ...
            </group>
```
That's the discovery of last month (thanks vfe)

**Review by Demesmaeker — COMMENTED** [review]
First part of the second pass, will continue tomorrow =] 
:duck:

File: addons/payment/models/payment_acquirer.py
Comment (chevalierv) [reply addons/payment/models/payment_acquirer.py:58]: An express checkout payment method allows collecting payment and address

**Review by Demesmaeker — COMMENTED** [review]
:duck:

File: addons/website_sale/controllers/main.py
Comment (chevalierv) [reply addons/website_sale/controllers/main.py:1227]:
```suggestion
        address.update(country_id=country,state_id=state)
```
Prefer using kwargs rather than call `dict.update` with a dict literal :) Good suggestion though ^^

Comment (Demesmaeker) [reply addons/website_sale/controllers/main.py]: Nice!! Thanks, I'll keep that in mind! =D

**Review by AntoineVDV — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**
